### PR TITLE
Adding cpp_info objects model for linking object files

### PR DESCRIPTION
**Pull request by czoido**
Changelog: Feature: Add objects attribute to the cpp_info so that we can add object files to the linker without having to add those mixed with linker flags.
Docs: https://github.com/conan-io/docs/pull/2243

For the moment it just works with CMakeDeps generator.

It's used like:

```
def package_info(self):
    self.cpp_info.components.objects = ["myobject.o"]
```
```
def package_info(self):
    self.cpp_info.components["say"].objects = ["mycomponent.o"]
```

Related with: 
- https://github.com/conan-io/conan/issues/8703
- https://github.com/conan-io/conan/issues/9438
- https://github.com/conan-io/conan/issues/5159
- https://github.com/conan-io/conan/pull/8740
- https://github.com/conan-io/conan/issues/8811

Related info:
- https://cmake.org/cmake/help/latest/prop_tgt/LINK_OPTIONS.html#prop_tgt:LINK_OPTIONS
- https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_LINK_OPTIONS.html
- https://cmake.org/cmake/help/latest/command/target_link_libraries.html#command:target_link_libraries
- https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_LINK_LIBRARIES.html
- https://cmake.org/cmake/help/latest/prop_tgt/LINK_LIBRARIES.html
- https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#object-libraries

### CODE REVIEW

**Review by memsharded — COMMENTED** [review]
One functional test that really links and execute an app with a couple of objects, for this feature please

**Comment by czoido** [comment]
> One functional test that really links and execute an app with a couple of objects, for this feature please

I'm on it, I already tested the feature with mimalloc: https://github.com/conan-io/conan-center-index/tree/master/recipes/mimalloc

File: conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py [hunk 1]
```diff
@@ -278,6 +278,60 @@ def package_info(self):
...
+    object_cpp = gen_function_cpp(name="myobject")
+    object_h = gen_function_h(name="myobject")
+    lib_cpp = gen_function_cpp(name="hello")
+    lib_h = gen_function_h(name="hello")
+    cmakelists = textwrap.dedent("""
+        cmake_minimum_required(VERSION 3.15)
+        project(HelloLib)
+        file(GLOB HEADERS *.h)
+        add_library(myobject OBJECT myobject.cpp)
+        add_library(hello hello.cpp hello.h)
```
Comment (memsharded) [inline conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py]: Why the "hello" library? Is not possible to have a package with just the objects? (I would say it is also an expected use case). I guess everything should work, but with CMake, who knows...

File: conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py [hunk 2]
```diff
@@ -278,6 +278,62 @@ def package_info(self):
     assert "MYVAR=>Like a Rolling Stone" in c.out
 
 
+@pytest.mark.tool_cmake
+def test_cpp_info_link_objects():
+    client = TestClient()
+
+    obj_ext = "obj" if platform.system() == "Windows" else "o"
+    cpp_info = {"objects": ["lib/myobject.{}".format(obj_ext)], "libs": ["hello"]}
```
Comment (lasote) [inline conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py]: I didn't know you can do this in a `GenConanfile` 😶‍🌫️.

**Review by lasote — COMMENTED** [review]
Pretty cool!

File: conan/tools/cmake/cmakedeps/templates/target_data.py [hunk 3]
```diff
@@ -66,6 +66,7 @@ def template(self):
               set({{ pkg_name }}_DEFINITIONS{{ config_suffix }} {{ global_cpp.defines }})
               set({{ pkg_name }}_SHARED_LINK_FLAGS{{ config_suffix }} {{ global_cpp.sharedlinkflags_list }})
               set({{ pkg_name }}_EXE_LINK_FLAGS{{ config_suffix }} {{ global_cpp.exelinkflags_list }})
+              set({{ pkg_name }}_OBJECTS{{ config_suffix }} {{ global_cpp.objects_list }})
```
Comment (memsharded) [inline conan/tools/cmake/cmakedeps/templates/target_data.py:69]: Only global? The cmake target_configuration seems to try to apply per-component ones, but seems per-component not defined here?

File: conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py [hunk 4]
```diff
@@ -85,3 +85,24 @@ def package_info(self):
...
+
+
+def test_cpp_info_component_objects():
+    client = TestClient()
+    conan_hello = textwrap.dedent("""
+        from conans import ConanFile
+        class Pkg(ConanFile):
+            settings = "os", "arch", "build_type"
+            def package_info(self):
+                self.cpp_info.components["say"].objects = ["/path/to/mycomponent.o"]
```
Comment (memsharded) [inline conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py]: It is not clear what the object should be, always an absolute path? How does it work with relative paths?

File: conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py [hunk 5]
```diff
@@ -278,6 +278,66 @@ def package_info(self):
     assert "MYVAR=>Like a Rolling Stone" in c.out
 
 
+@pytest.mark.tool_cmake
+def test_cpp_info_link_objects():
+    client = TestClient()
+    obj_ext = "obj" if platform.system() == "Windows" else "o"
+    cpp_info = {"objects": [os.path.join("lib", "myobject.{}".format(obj_ext))], "libs": ["hello"]}
```
Comment (memsharded) [inline conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py]: This path is relative, but seems to work (see note below)

File: conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py (hunk 4 above)
Comment (czoido) [reply conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py]: Sorry, this is a mistake, it should be a relative path

File: conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py (hunk 5 above)
Comment (czoido) [reply conans/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps.py]: Yes, it should always be a relative path, I have changed the others that were mistaken

**Review by memsharded — APPROVED** [review]

**Review by lasote — APPROVED** [review]

**Pull request merged**